Build: Change package build step to async flow

aduth opened the pull request:
This pull request seeks to adapt the package build script to perform builds in parallel, rather than in sequence. In practice, this unfortunately does not appear to make a significant difference, reducing the build time for me from about 12s to 10s. It seems that the style build takes a significant part of the time of package build.

Noting also that this is not a total conversion. `getPackages` should be made to be asynchronous, but is touched by other scripts as well. As such, it was decided to be left as a separate task.

**Testing instructions:**

There should be no regressions in the build of packages.

youknowriad commented on bin/packages/build.js:
```diff
@@ -132,34 +148,42 @@ function buildFileFor( file, silent, environment ) {
+	return Promise.all( [
+		...files.map( ( file ) => buildFile( file, true ) ),
+		buildPackageStyleIfApplicable(),
+	] ).then( () => {
```
This could be replaced with `await` ?

aduth replied:
> This could be replaced with `await` ?

Yep, updated in rebased 9feaf4a.

youknowriad reviewed (approved):
LGTM 👍

gziolo commented:
This one needs rebase and some adjustments after changes introduced in #8187 for RTL support for CSS.

aduth committed:
Build: Change package build step to async flow

aduth commented:
I've finally brought this back up to date. Unfortunately it still doesn't seem to have much of an overall impact.

aduth commented on bin/packages/build.js:
```diff
@@ -132,34 +148,42 @@ function buildFileFor( file, silent, environment ) {
-		.forEach( buildPackage );
-	process.stdout.write( '\n' );
+
+	Promise.all( [
```
> Shouldn't it be:
> 
> ```js
> Promise.all( getPackages().map( buildPackage ) )...
> ```
> 
> ?

Updated in rebased 9feaf4a.

gziolo commented:
Feel free to merge when Travis is happy. It was already reviewed :)

> I've finally brought this back up to date. Unfortunately it still doesn't seem to have much of an overall impact.

@noisysocks has PR opened which adds caching, maybe it will have an impact together.

aduth commented:
I have two of my own remaining reservations:

- I'm not sure about `postcss().process` being synchronous as it was refactored here. The function is a bit confusing, where it [returns a "LazyResult"](http://api.postcss.org/Processor.html#process) object which has both an [evaluated `css` property](http://api.postcss.org/LazyResult.html#css) but also is [`then`-able](http://api.postcss.org/LazyResult.html#then). By the documentation of `then`, I'm inclined to think that it must be used if needing to account for asynchronous postcss plugins.
- I'm not sure about unbounded concurrency for the build. It's probably not much of an issue to build our current set of modules in parallel, but it might be worth considering something like [`p-limit`](https://www.npmjs.com/package/p-limit) to batch a set maximum number of builds to occur at a time.

aduth commented:
Ah, I guess I should have read those links more closely, as it answers my question quite concretely 😅 

>This is why this method is only for debug purpose, you should always use LazyResult#then.

http://api.postcss.org/LazyResult.html#css

aduth committed:
Build: Trigger LazyResult#then for postcss.process

See: http://api.postcss.org/LazyResult.html#css

>This is why this method is only for debug purpose, you should always use LazyResult#then.

aduth committed:
Build: Consistently use css property for stringified result

Effectively unchanged, since stringification of the LazyResult object is aliased to the css property.

See: http://api.postcss.org/LazyResult.html#toString

aduth commented:
`LazyResult` is now used as a thenable as of b6e7084.

I'm fine in the initial approach to have unbounded parallelization. This can be revisited if it's shown to be problematic.

aduth commented:
There's a few issues here impacting (a) sourcemaps, (b) `npm run dev` regeneration, and (c) a possible race condition with `writeFile`.

I've proposed a revert at #13195 

I have an in-progress branch with some partial fixes at [`fix/mkdirp-source-maps`](https://github.com/WordPress/gutenberg/compare/fix/mkdirp-source-maps).

I'll leave a few follow-up comments here for more specific problems.

aduth commented on bin/packages/build.js:
```diff
@@ -87,39 +97,50 @@ function buildFiles( files ) {
-	buildPaths.jsFiles.forEach( buildJsFile );
-	buildPaths.scssPackagePaths.forEach( buildPackageScss );
+	return Promise.all( [
+		...buildPaths.jsFiles.map( buildJsFile ),
```
I was wrong in thinking that by the previous implementation using `forEach`, these were arrays, when in fact they're of type `Set`. Thus, rebuilds fail with an error:

```
[1] /Users/andrew/Documents/Code/vvv/www/editor/htdocs/wp-content/plugins/gutenberg/bin/packages/build.js:100
[1] 		...buildPaths.jsFiles.map( buildJsFile ),
```

Two alternatives are:

- Don't track them as a `Set`, rather as an array and guarantee uniqueness before pushing to array.
- Convert `Set` to array at this line before the map, i.e. `...[ ...buildPaths.jsFiles ].map( buildJsFile ),`

aduth commented on bin/packages/build.js:
```diff
@@ -161,17 +177,17 @@ function buildScssFile( styleFile ) {
-	fs.writeFileSync( destPath, transformed.code + '\n//# sourceMappingURL=' + path.basename( destPath ) + '.map' );
+	await mkdirp( path.dirname( destPath ) );
+	const transformed = await transformFile( file, babelOptions );
+	writeFile( destPath + '.map', JSON.stringify( transformed.map ) );
```
These are asynchronous tasks and should have `await`, preferably in parallel as `await Promise.all` between writing the map file and the original source.

aduth commented on bin/packages/build.js:
```diff
@@ -161,17 +177,17 @@ function buildScssFile( styleFile ) {
-	const transformed = babel.transformFileSync( file, babelOptions );
-	fs.writeFileSync( destPath + '.map', JSON.stringify( transformed.map ) );
-	fs.writeFileSync( destPath, transformed.code + '\n//# sourceMappingURL=' + path.basename( destPath ) + '.map' );
+	await mkdirp( path.dirname( destPath ) );
```
I'm seeing varied results in attributing this to be a cause, but in particular I was seeing some improvement by switching this back to its synchronous form.

Possibly related: https://github.com/substack/node-mkdirp/issues/154

Viable substitute module: https://www.npmjs.com/package/make-dir

talldan commented:
I had a similar attempt at this over christmas (didn't realise this PR existed). I also didn't see any real improvement in build times, so didn't bother tidying things up:
https://github.com/WordPress/gutenberg/pull/13104

It'll be worth keeping an eye on node 10's promise based filesystem apis. They're currently marked as experimental, hopefully in a not too distant future version they'll become stable:
https://nodejs.org/docs/latest-v10.x/api/fs.html#fs_fs_promises_api
